### DISCUSSION TITLE
Github as Data Source - EMU Instance

### DISCUSSION THREAD

**Discussion by Mailstorm-ctrl** [netbox-community/netbox#14941] [Q&A] [answered]
Is it possible to use github as a data source when you have an EMU instance? EMU instances don't allow the username/password auth as all authentication is done through the configured Identity Provider.  Regular github also doesn't allow it so i'm not really sure how the github data source is supposed to work when it uses tokens for authentication

**Comment by markkuleinio** [comment]
I don't know about EMU, but if it supports SSH keys, you may be able to use them.

Based on https://docs.netbox.dev/en/stable/configuration/system/#git_path NetBox uses normal git command and that supports using SSH keys.

**Reply by Mailstorm-ctrl** [reply]
Neat. This could be it. I'm fairly new to github but it seems like I can make deploy keys which is what I'm guessing you're talking about. Looks like I just make a key in github, import it into the vm netbox is on and bam?

**Reply by markkuleinio** [reply]
You need to create the key on your host, then copy the public key in the deploy keys of the repo.

But, I just tried this (saved the key to /home/netbox/.ssh/id_ed25519 and so on, to get git to use the key by default, the usual stuff how I do elsewhere to use git without interactive authentication) and got error:

**SyncError('Fetching remote data failed (HTTPUnauthorized): No valid credentials provided')**

So I'm not sure if that will work at all in this case. Maybe this is tailored to use username+password auth only?

Anyone else using SSH keys with git in NetBox data sources?

**Reply by Mailstorm-ctrl** [reply]
Bummer. However you did answer an unrelated question about another project I'm working on. Didn't realize I could use SSH to read/write to a repo.

**Reply by markkuleinio** [reply]
Actually I used incorrect URL above 😂

But it still doesn't work. Just a minute, I'll try something

**Reply by abhi1693** [reply]
data sources use `dulwich` package rather than `git` binary

**Reply by markkuleinio** [reply]
Why is this then:

> GIT_PATH
> Default: git
> The system path to the git executable, used by the synchronization backend for remote git repositories.

**Reply by markkuleinio** [reply]
**SyncError('Fetching remote data failed (HangupException): git@github.com: Permission denied (publickey).\r')**

is my error now, so it does not use the user's ed25519 key (which is tested and working, and properly chmod'ed)

**Reply by abhi1693** [reply]
I never looked up the source code for GIT_PATH until now and it seems like it's not used anywhere. It's an old remnant when https://github.com/netbox-community/netbox/pull/12274 was merged.

**Reply by markkuleinio** [reply]
Maybe someone is able to tell how to use SSH keys with `dulwich` then

**Reply by Mailstorm-ctrl** [reply]
I looked around and it seems like using SSH keys will require a DB modification to store the path of an SSH key, in addition an extra field will be needed when making a git datasource. `Dulwich` is it's own thing and seems to ignore anything outside of its configuration.

I could be wrong, but that's what it really seems like after playing around.

**Reply by Mailstorm-ctrl** [reply]
> **SyncError('Fetching remote data failed (HangupException): [git@github.com](mailto:git@github.com): Permission denied (publickey).\r')**
> 
> is my error now, so it does not use the user's ed25519 key (which is tested and working, and properly chmod'ed)

I'm revisiting this since the Issue I opened up changed status. Do you remember if you acted as the netbox user and attempted to ssh into github? Seems like there's more to it

**Reply by Mailstorm-ctrl** [reply, marked as answer]
Howdy ho I got this working with no code changes
https://github.com/netbox-community/netbox/issues/15394
